Code/GraphMol/Descriptors/BCUT.cpp

```diff
@@ -0,0 +1,136 @@
+//
+//  Copyright (C) 2020 Brian P. Kelley
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#ifdef RDK_HAS_EIGEN3
+#include "BCUT.h"
+#include "Crippen.h"
+#include <Eigen/Dense>
+#include <GraphMol/RDKitBase.h>
+#include "GraphMol/PartialCharges/GasteigerCharges.h"
+#include "GraphMol/PartialCharges/GasteigerParams.h"
+#include <RDGeneral/types.h>
+
+namespace RDKit {
+namespace Descriptors {
+  // diagonal elements are a property (atomic num, charge, etc)
+  // off diagonal are 1/sqrt(bond_order)
+  //  Original burden matrix was .1, .2, .3, .15 for single,double,triple or aromatic
+  //  all other elements are .001
+namespace {
+std::unique_ptr<Eigen::MatrixXd>  make_burden(const ROMol &m) {
+  auto num_atoms = m.getNumAtoms();
+  std::unique_ptr<Eigen::MatrixXd> burden(
+	 new Eigen::MatrixXd(num_atoms, num_atoms));
+    
+  for(unsigned int i=0;i<num_atoms; ++i) {
+    for(unsigned int j=0;j<num_atoms; ++j) {
+      (*burden)(i,j) = (*burden)(j,i) = 0.001;
+    }
+  }
+  
+  for(auto &bond : m.bonds()) {
+    unsigned int i = bond->getBeginAtomIdx();
+    unsigned int j = bond->getEndAtomIdx();
+    double score = 0.0;
+    switch(bond->getBondType()) {
+    case Bond::AROMATIC:
+      // score = 0.15; orig burden
+      score = 0.8164965809277261; // 1/sqrt(1.5)
+      break;
+    case Bond::SINGLE:
+      // score = 0.1;
+      score = 1.0; // 1/sqrt(1.0)
+      break;
+    case Bond::DOUBLE:
+      // score = 0.2;
+      score = 0.7071067811865475; // 1/sqrt(2.0)
+      break;
+    case Bond::TRIPLE:
+      // score = 0.3;	
+      score = 0.5773502691896258; // 1/sqrt(3);
+      break;
+    default:
+      CHECK_INVARIANT(0, "Bond order must be Single, Double, Triple or Aromatic");
+    }
+    (*burden)(i,j) = (*burden)(j,i) = score;
+  }    
+  return burden;
+}
+  
+std::pair<double,double> BCUT2D(std::unique_ptr<Eigen::MatrixXd> &burden,
+				const std::vector<double> &atom_props) {
+  for(unsigned int i=0; i<atom_props.size(); ++i) {
```

Comment:
Probably a good idea to have some preconditions on this. Something like:
```
PRECONDITION(burden.rows()==burden.cols()==atom_props.size(),"bad matrix shape");
```

Reply:
I do the checks before calling this function (it isn't user accessible) so I think it's ok.